File: go/algorithms/sort/merge_sort.go
package sort

func mergeSort(array []int) []int {
	length := len(array)

	if length == 1 { // базовый случай рекурсии
		return array
	}

	mid := length // 2

	// запускаем сортировку рекурсивно на левой половине
	left := mergeSort(array[:mid])

	// запускаем сортировку рекурсивно на правой половине
	right := mergeSort(array[mid:])

	// заводим массив для результата сортировки
	result := make([]int, length)

	// сливаем результаты
	l, r, k := 0, 0, 0
	for l < len(left) && r < len(right) {
		// выбираем, из какого массива забрать минимальный элемент
		if left[l] <= right[r] {
			result[k] = left[l]
			l++
		} else {
			result[k] = right[r]
			r++
		}
		k++
	}

	// Если один массив закончился раньше, чем второй, то
	// переносим оставшиеся элементы второго массива в результирующий
	for l < len(left) {
		result[k] = left[l] // перенеси оставшиеся элементы left в result
		l++
		k++
	}
	for r < len(right) {
		result[k] = right[r] // перенеси оставшиеся элементы right в result
		r++
		k++
	}

	return result
}

File: python/src/algorithms/sort/merge_sort.py
def merge_sort(array: list[int]) -> list[int]:
    length = len(array)

    if length == 1:  # базовый случай рекурсии
        return array

    mid = length // 2

    # запускаем сортировку рекурсивно на левой половине
    left = merge_sort(array[:mid])

    # запускаем сортировку рекурсивно на правой половине
    right = merge_sort(array[mid:])

    # заводим массив для результата сортировки
    result = [0] * length

    # сливаем результаты
    lf, rt, k = 0, 0, 0
    while lf < len(left) and rt < len(right):
        # выбираем, из какого массива забрать минимальный элемент
        if left[lf] <= right[rt]:
            result[k] = left[lf]
            lf += 1
        else:
            result[k] = right[rt]
            rt += 1
        k += 1

    # Если один массив закончился раньше, чем второй, то
    # переносим оставшиеся элементы второго массива в результирующий
    while lf < len(left):
        result[k] = left[lf]  # перенеси оставшиеся элементы left в result
        lf += 1
        k += 1
    while rt < len(right):
        result[k] = right[rt]  # перенеси оставшиеся элементы right в result
        rt += 1
        k += 1

    return result


if __name__ == "__main__":
    array = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]
    sorted_array = merge_sort(array)
    print(sorted_array)
